File: src/space/cherryband/ari/data/SlobDescriptor.java
package space.cherryband.ari.data;

import android.content.Context;
import android.net.Uri;
import android.os.ParcelFileDescriptor;
import android.util.Log;

import java.io.FileInputStream;
import java.util.HashMap;
import java.util.Map;

import itkach.slob.Slob;


public class SlobDescriptor extends BaseDescriptor {

    private final static transient String TAG = SlobDescriptor.class.getSimpleName();

    public String path;
    public Map<String, String> tags = new HashMap<>();
    public boolean active = true;
    public long priority;
    public long blobCount;
    public String error;
    public boolean expandDetail = false;
    private transient ParcelFileDescriptor fileDescriptor;

    void update(Slob s) {
        this.id = s.getId().toString();
        this.path = s.fileURI;
        this.tags = s.getTags();
        this.blobCount = s.getBlobCount();
        this.error = null;
    }

    public Slob load(final Context context) {
        Slob slob = null;
        //File f = new File(path);

        try {
            //slob = new Slob(f);
            final Uri uri = Uri.parse(path);
            //must hold on to ParcelFileDescriptor,
            //otherwise it gets garbage collected and trashes underlying file descriptor
            fileDescriptor = context.getContentResolver().openFileDescriptor(uri, "r");
            FileInputStream fileInputStream = new FileInputStream(fileDescriptor.getFileDescriptor());
            slob = new Slob(fileInputStream.getChannel(), path);
            this.update(slob);
        } catch (Exception e) {
            Log.e(TAG, "Error while opening " + this.path, e);
            error = e.getMessage();
            if (Log.isLoggable(TAG, Log.DEBUG)) {
                Log.d(TAG, "Error while opening " + this.path, e);
            }
            expandDetail = true;
            active = false;
        }
        return slob;
    }

    public String getLabel() {
        String label = tags.get("label");
        if (label == null || label.trim().length() == 0) {
            label = "???";
        }
        return label;
    }

//    static SlobDescriptor fromFile(File file) {
//        SlobDescriptor s = new SlobDescriptor();
//        s.path = file.getAbsolutePath();
//        s.load();
//        return s;
//    }

    public static SlobDescriptor fromUri(Context context, String uri) {
        SlobDescriptor s = new SlobDescriptor();
        s.path = uri;
        s.load(context);
        return s;
    }
}

File: src/space/cherryband/ari/data/SlobDescriptor.kt
package space.cherryband.ari.data

import android.content.Context
import android.net.Uri
import android.os.ParcelFileDescriptor
import android.util.Log
import itkach.slob.Slob
import java.io.FileInputStream

class SlobDescriptor : BaseDescriptor() {
    var path: String? = null
    var tags: Map<String, String> = HashMap()
    var active = true
    var priority: Long = 0
    var blobCount: Long = 0
    var error: String? = null
    var expandDetail = false

    @Transient
    private var fileDescriptor: ParcelFileDescriptor? = null
    private fun update(s: Slob) {
        id = s.id.toString()
        path = s.fileURI
        tags = s.tags
        blobCount = s.blobCount
        error = null
    }

    fun load(context: Context): Slob? {
        var slob: Slob? = null
        //File f = new File(path);
        try {
            //slob = new Slob(f);
            val uri = Uri.parse(path)
            //must hold on to ParcelFileDescriptor,
            //otherwise it gets garbage collected and trashes underlying file descriptor
            fileDescriptor = context.contentResolver.openFileDescriptor(uri, "r")
            val fileInputStream = FileInputStream(fileDescriptor!!.fileDescriptor)
            slob = Slob(fileInputStream.channel, path)
            update(slob)
        } catch (e: Exception) {
            Log.e(TAG, "Error while opening $path", e)
            error = e.message
            if (Log.isLoggable(TAG, Log.DEBUG)) {
                Log.d(TAG, "Error while opening $path", e)
            }
            expandDetail = true
            active = false
        }
        return slob
    }

    val label: String
        get() {
            var label = tags["label"]
            if (label == null || label.trim { it <= ' ' }.isEmpty()) {
                label = "???"
            }
            return label
        }

    companion object {
        @Transient
        private val TAG = SlobDescriptor::class.java.simpleName

    //  fun fromFile(file: File): SlobDescriptor {
    //      val s = SlobDescriptor()
    //      s.path = file.absolutePath
    //      s.load()
    //      return s
    //  }
        fun fromUri(context: Context, uri: String?): SlobDescriptor {
            val s = SlobDescriptor()
            s.path = uri
            s.load(context)
            return s
        }
    }
}